src/v/cluster/self_test/cloudcheck.cc

```diff
@@ -0,0 +1,410 @@
+    result.name = result_name;
+    result.info = result_info;
+    result.test_type = result_test_type;
+
+    if (!object_list) {
+        result.error = "Failed to list objects in cloud storage";
+    }
+
+    co_return std::make_pair(std::move(object_list), std::move(result));
+}
+
+ss::future<std::pair<std::optional<iobuf>, self_test_result>>
+cloudcheck::verify_download(
+  const cloud_storage_clients::bucket_name& bucket,
+  const std::optional<cloud_storage_clients::object_key>& key) {
+    const auto& result_name = _opts.name;
+    static const auto result_info = "download";
+    static const auto result_test_type = "cloud storage";
+
+    if (_cancelled) {
+        co_return std::make_pair(
+          std::nullopt,
+          self_test_result{
+            .name = result_name,
+            .info = result_info,
+            .test_type = result_test_type,
+            .warning = "Run was manually cancelled."});
+    }
+
+    if (!_remote_read_enabled) {
+        co_return std::make_pair(
+          std::nullopt,
+          self_test_result{
+            .name = result_name,
+            .info = result_info,
+            .test_type = result_test_type,
+            .error = "Remote read is not enabled for this cluster."});
+    }
+
+    if (!key) {
+        co_return std::make_pair(
+          std::nullopt,
+          self_test_result{
+            .name = result_name,
+            .info = result_info,
+            .test_type = result_test_type,
+            .error = "Failed to download from cloud storage (download key "
+                     "provided is std::nullopt)."});
+    }
+
+    iobuf download_payload;
+    cloud_storage::download_request download_request = make_download_request(
+      bucket, key.value(), std::ref(download_payload));
+
+    const auto start = ss::lowres_clock::now();
+    const cloud_storage::download_result download_result
+      = co_await _cloud_storage_api.local().download_object(
+        std::move(download_request));
+    const auto end = ss::lowres_clock::now();
+
+    const auto duration = end - start;
+    auto result = self_test_result{.duration = duration};
+    result.name = result_name;
+    result.info = result_info;
+    result.test_type = result_test_type;
+
+    std::optional<iobuf> result_payload;
+    switch (download_result) {
+    case cloud_storage::download_result::success:
+        result_payload = std::move(download_payload);
+        break;
+    case cloud_storage::download_result::timedout:
+        [[fallthrough]];
+    case cloud_storage::download_result::failed:
+        [[fallthrough]];
+    case cloud_storage::download_result::notfound:
+        result.error = "Failed to download from cloud storage.";
+        break;
+    }
+
+    co_return std::make_pair(std::move(result_payload), std::move(result));
+}
+
+ss::future<self_test_result> cloudcheck::verify_delete(
+  const cloud_storage_clients::bucket_name& bucket,
+  const cloud_storage_clients::object_key& key) {
```

Comment:
general best practice unless performance sensitive is to have coroutines take parameters by value.